tests/python/relax/test_dataflow_rewriter.py

```diff
@@ -0,0 +1,1370 @@
+
+
+def test_rewrite_defined_by_ir_module():
+    @R.rewriter
+    class Rewriter:
+        @R.function
+        def pattern(A: R.Tensor([16], "float32"), B: R.Tensor([16], "float32")):
+            C = R.add(A, B)
+            return C
+
+        @R.function
+        def replacement(A: R.Tensor([16], "float32"), B: R.Tensor([16], "float32")):
+            C = R.call_pure_packed(
+                "my_optimized_add_impl", A, B, sinfo_args=R.Tensor([16], "float32")
+            )
+            return C
+
+    @R.function
+    def before(x: R.Tensor([32], "float32")):
+        R.func_attr({"global_symbol": "main"})
+        split = R.split(x, 2)
+        lhs = split[0]
+        rhs = split[1]
+        out = lhs + rhs
+        return out
+
+    @R.function
+    def expected(x: R.Tensor([32], "float32")):
+        R.func_attr({"global_symbol": "main"})
+        split = R.split(x, 2)
+        lhs = split[0]
+        rhs = split[1]
+        out = R.call_pure_packed(
+            "my_optimized_add_impl", lhs, rhs, sinfo_args=R.Tensor([16], "float32")
+        )
+        return out
+
+    after = Rewriter(before)
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_missing_pattern_raises_error():
+    """The rewriter must define a pattern to be matched"""
+
+    with pytest.raises(KeyError, match="pattern"):
+
+        @R.rewriter
+        class Rewriter:
+            @R.function
+            def replacement():
+                return R.tuple()
+
+
+def test_incorrect_function_type_of_pattern_raises_error():
+    """The rewriter's pattern must be a Relax function"""
+
+    with pytest.raises(TypeError, match="pattern"):
+
+        @R.rewriter
+        class Rewriter:
+            @T.prim_func
+            def pattern():
+                pass
+
+            @R.function
+            def replacement():
+                return R.tuple()
+
+
+def test_missing_replacement_raises_error():
+    """The rewriter must define a replacement"""
+
+    with pytest.raises(KeyError, match="replacement"):
+
+        @R.rewriter
+        class Rewriter:
+            @R.function
+            def pattern():
+                return R.tuple()
+
+
+def test_incorrect_function_type_of_replacement_raises_error():
+    """The rewriter's replacement must be a Relax function"""
+
+    with pytest.raises(TypeError, match="replacement"):
+
+        @R.rewriter
+        class Rewriter:
+            @R.function
+            def pattern():
+                return R.tuple()
+
+            @T.prim_func
+            def replacement():
+                pass
+
+
+def test_mismatch_of_static_shapes_raises_error():
+    """The pattern and replacement must accept the same shapes"""
+
+    with pytest.raises(ValueError, match="must have the same signature"):
+
+        @R.rewriter
+        class Rewriter:
+            @R.function
+            def pattern(A: R.Tensor([32])):
+                return A
+
+            @R.function
+            def replacement(A: R.Tensor([16])):
```

Comment:
Out of curiosity, do we catch these errors during parsing? or do we defer until we run the replacement?

Reply:
These errors are caught while converting the `IRModule`, as part of the `ExprRewriter::FromModule` function, before the `@R.rewriter` annotator returns the rewriter.  So, technically it occurs after parsing (since the IRModule is fully defined) but before replacement (since the `ExprRewriter` is validated on construction).